### PR TITLE
Mandelbrot videogame

### PR DESCRIPTION
**Pull request by BrunoPlatero**
Now you can travel in the mandelbrot universe

Used SDL to open a window and handle user input

### CODE REVIEW

File: src/fractal.c
Comment (ManuCanedo) [inline src/fractal.c]: Space missing before the comment

File: src/fractal.c
Comment (ManuCanedo) [inline src/fractal.c]: You can stop looping if ab(Z) is bigger than the 'radius_2d'. Imagine in iteration 2, Z^2 is already bigger than 4; you will still loop 252 times, performing maths but never setting the pixel's colour (you already set it).

Comment (ManuCanedo) [reply src/fractal.c]: You are doing this:
```
while it <= max_its:
    Z = Z * Z + C
    if abs(Z) > 4:
        pixel = ITERATION_BASED_COLOUR
    if it == max_its:
        pixel = BLACK
    # keep going even though you already know the colour for that pixel
````
It would be best if you were doing this:
````
while it < max_its && abs(Z) < 4:
    Z = Z * Z + C
    it++

if if == max_it: # The while stopped due to the `it < max_its` check
    pixel = BLACK
else: # The while stopped due to the `abs(Z) < 4` check
    pixel = ITERATION_BASED_COLOUR
```

Comment (ManuCanedo) [reply src/fractal.c]: Also, please rename radius_2d with something more meaningful.

File: src/fractal.c
Comment (ManuCanedo) [inline src/fractal.c]: Don't commit to the main branch commented code.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: Maintain alphabetical order

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: Rename type to something more meaningful. This represents a type that contains the number of elements in the memory where we store the pixels.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: Tabs are still broken here.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: Indeed, if an error happens you should return from main here (and on all the other commented returns down below). I let you reason what is the appropriate return in these cases.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: You need to fully debug these movement logics before this can be merged to main.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: Don't commit commented code.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: Too many spaces. You should review the code format before making a commit. A commit should be code that is polished 100%.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: More commented code.

File: Makefile
Comment (ManuCanedo) [inline Makefile:42]: Add a comment here stating that this Makefile target requires bear installed and how someone can install it.

**Review by ManuCanedo — CHANGES_REQUESTED** [review]
Your progress is looking neat! Bear in mind that the code in the main branch has to be very polished and debugged. You shouldn't be committing code that you know contains some bugs (the movement code) or commented functionalities.

File: Makefile
Comment (ManuCanedo) [inline Makefile]: `-lSDL2` is a linker flag. It doesn't need to be added to any compile commands. The flag `-fopenmp` is an exception because it also tells the compiler where to find the openmp header (include flags are compiler flags).

A summary of compile flags vs link flags would be this:
Compile flags (CFLAGS) are all the flags required to compile .c files into .o files. Among others, they include:

1. Include paths (where to find .h files). `-Ipath/to/headers`
2. Optimization flags. `-O3`
3. Warning controlling flags. `-Wall -Werror -Wextra`. 

Linker flags (LDFLAGS) are all the flags required to link all .o files into a binary. Among others, they include:

1. Libraries. `-lSDL2`
2. Library paths (where to find the linked libraries). `-Lpath/to/lib`
3. Linker optimizations. `-flto`  

Other "special" flags are linker and compiler flags, such as `-fopenmp`. It is a compile flag because it tells the compiler to find the OpenMP headers. It is a linker flag because it tells the compiler to link against the openmp library and where to find it.

**Comment by BrunoPlatero** [comment]
king

File: src/bmp.c
Comment (ManuCanedo) [inline src/bmp.c]: It would help if you tried never to use global variables. These two are used only in the write_bmp function, so they both should be in that function. Could you read the further comment below?

File: src/bmp.c
Comment (ManuCanedo) [inline src/bmp.c:14]: You can declare screenshots_count here. You need it to "survive" between function calls so that you can declare the variable 'static'. ChatGPT explains it:

What does static mean, in a variable declaration inside a function, in C?
In C, when a variable is declared static inside a function, it retains its value between function calls. This means that the variable's value is preserved between function calls and is not reinitialized each time the function is called.

For example:
```
#include <stdio.h>

void increment() {
  static int counter = 0;
  counter++;
  printf("Counter: %d\n", counter);
}

int main() {
  increment();  // prints "Counter: 1"
  increment();  // prints "Counter: 2"
  increment();  // prints "Counter: 3"
  return 0;
}
```
In this code, the function increment increments a counter variable each time it is called. The counter variable is declared static, so its value is preserved between function calls. The first time the function is called, the counter is initialized to 0 and then incremented to 1. The second time the function is called, the counter is already at 1, so it is incremented to 2. This process continues each time the function is called.

In contrast, if the counter variable were not declared static, it would be reinitialized to 0 each time the function was called, and the output would always be "Counter: 1".

Comment (ManuCanedo) [reply src/bmp.c:14]: The other variable, new_filename, doesn't need to be static as it doesn't require preserving its value between calls. That can be just a standard array declared inside the function.

File: src/fractal.c
Comment (ManuCanedo) [inline src/fractal.c]: You are doing a modulus with 256 here. You should do % 256 in the lines above and below as well. That gives you a number between 0-255.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: This is a MEMORY_ARRAY_SIZE. Because it's a number that represents the number of elements in an array. A type in the standard library exists specifically for this use: size_t.

It would help if you used size_t when you want to represent several bytes/elements in an array, and you can remove this typedef altogether.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: If you want to define motion as a hardcoded global constant of your program, you should use a macro. The same you did for COLUMNS and ROWS.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: size_t data_size = ...

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: This ptr should have another name now that it just holds the frame memory.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c:113]: A future improvement would be to create a function that takes a world* and scales it a certain amount, to avoid code repetition.
```
void scale_world(struct world_coordinates* world, double percent) {
    double center_x = (world->set_right + world->set_left) / 2;
    double center_y = (world->set_up + world->set_down) / 2;

    // Translate the object, so its centre is at the origin
    world->set_left -= center_x;
    world->set_right -= center_x;
    world->set_down -= center_y;
    world->set_up -= center_y;

    // Scale the object
    world->set_left *= 1 + percent;
    world->set_right *= 1 + percent;
    world->set_down *= 1 + percent;
    world->set_up *= 1 + percent;

    // Translate the object back to its original position
    world->set_left += center_x;
    world->set_right += center_x;
    world->set_down += center_y;
    world->set_up += center_y;
}
```
You can call it with positive or negative percents to zoom in or out.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c]: A future improvement would be to create a function that takes a world* and translates it a certain amount to avoid code repetition. I will implement a draft below:
```
void translate_world(struct world_coordinates* world, double horizontal_percent, double vertical_percent) {
        long double horizontal_increment = (world->set_right - world->set_left) * horizontal_percent;
        world->set_right += horizontal_increment;
        world->set_left += horizontal_increment;

        long double vertical_increment = (world->set_up - world->set_down) * vertical_percent;
        world->set_up += vertical_increment;
        world->set_down += vertical_increment;
}
```
You can call it with positive percent to move the world right or up and negative percent to push it down or left.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c:94]: A future improvement could be to replace this if-else if with a switch statement:
```
switch (event.type) {
case SDL_QUIT:
    // handle quit
    break;
case SDLK_a:
    // handle a
    break;
...
}
```

File: src/main.c
Comment (ManuCanedo) [inline src/main.c:178]: You should free the malloc'd memory as well.

**Review by ManuCanedo — CHANGES_REQUESTED** [review]

**Comment by ManuCanedo** [comment]
This is looking amazing! Some minor comments on minor fixes and some future improvements I propose.

File: src/bmp.c
Comment (BrunoPlatero) [reply src/bmp.c]: done

File: src/fractal.c
Comment (BrunoPlatero) [reply src/fractal.c]: done

File: src/main.c
Comment (BrunoPlatero) [reply src/main.c]: done

File: src/main.c
Comment (BrunoPlatero) [reply src/main.c]: done

File: src/bmp.c
Comment (ManuCanedo) [inline src/bmp.c]: You missed moving this into the write_bmp function.

File: src/bmp.c
Comment (ManuCanedo) [inline src/bmp.c]: You forgot to make this static. This will make every screenshot override the previous.

File: src/main.c
Comment (ManuCanedo) [inline src/main.c:17]: You should move this user welcome code into a function. Good names could be `print_user_welcome` or something along those lines.

**Review by ManuCanedo — CHANGES_REQUESTED** [review]

File: src/bmp.c
Comment (BrunoPlatero) [reply src/bmp.c]: done

**Review by ManuCanedo — APPROVED** [review]
lgtm!

**Pull request merged**
